src/BotPlutusInterface/Balance.hs

```diff
@@ -281,17 +282,26 @@ balanceTxIns utxos tx = do
  (suboptimally we just pick a random utxo from the tx inputs)
 -}
 addTxCollaterals :: Map TxOutRef TxOut -> Tx -> Either Text Tx
-addTxCollaterals utxos tx = do
-  let txIns = mapMaybe (rightToMaybe . txOutToTxIn) $ Map.toList $ filterAdaOnly utxos
-  txIn <- findPubKeyTxIn txIns
-  pure $ tx {txCollateral = Set.singleton txIn}
+addTxCollaterals utxos tx =
+  if not $ usesScripts tx
+    then Right tx
+    else do
+      let txIns = mapMaybe (rightToMaybe . txOutToTxIn) $ Map.toList $ filterAdaOnly utxos
+      txIn <- findPubKeyTxIn txIns
+      pure $ tx {txCollateral = Set.singleton txIn}
   where
     findPubKeyTxIn = \case
       x@(TxIn _ (Just ConsumePublicKeyAddress)) : _ -> Right x
       x@(TxIn _ Nothing) : _ -> Right x
       _ : xs -> findPubKeyTxIn xs
       _ -> Left "There are no utxos to be used as collateral"
     filterAdaOnly = Map.filter (isAdaOnly . txOutValue)
+    usesScripts Tx {txInputs, txMintScripts} =
+      not (null txMintScripts)
+        || isJust
```

Comment:
Can we use `any` over `isJust` and `find`?

Reply:
ah hah, I knew there'd be a better way to do this, thanks! Updated.